In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import yaml
%matplotlib inline

try:
    with open("../config.yaml", "r") as file:
        config = yaml.safe_load(file)
except:
    print("Yaml configuration file not found!")

In [2]:
config['data']['raw']['file_footprints']

'../data/raw/df_footprints.csv'

In [ ]:
#df.to_csv("plath_to_destination_file", index=False)

In [3]:
df_final_experiment = pd.read_csv(config['data']['raw']['file_experiment_clients'])

In [4]:
df_footprints = pd.read_csv(config['data']['raw']['file_footprints'])
df_footprints = df_footprints.drop(columns=['Unnamed: 0'])

df_footprints

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
755400    9668240   388766751_9038881013   922267647_3096648104_968866   
755401    9668240   388766751_9038881013   922267647_3096648104_968866   
755402    9668240   388766751_9038881013   922267647_3096648104_968866   
755403    9668240   388766751_9038881013   922267647_3096648104_968866   
755404     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
755400        start  2017-05-24 18:46:10  
755401        start  2017-05-24 18:45:29  
755402       step_1  2017-05-24 18:44:51  
755403        start  2017-05-24 18:44:34  
755404        start  2017-06-03 12:17:09  

[755405 rows x 5 columns]

In [5]:
df_footprints.columns


Index(['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time'], dtype='object')

In [6]:
df_footprints.isnull().sum()

client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

In [7]:
total_duplicates = df_footprints.duplicated().sum()
print(total_duplicates)

10764


In [8]:
df_footprints[df_footprints.duplicated()].sort_values(by="client_id")


client_id             visitor_id                      visit_id  \
631356       1531  934069404_34543643308  110350144_67985219545_904677   
631354       1531  934069404_34543643308  110350144_67985219545_904677   
639519       2078  585735301_21309149782  900598259_99565669243_552843   
639522       2078  585735301_21309149782  900598259_99565669243_552843   
639530       2078  585735301_21309149782  900598259_99565669243_552843   
...           ...                    ...                           ...   
638238    9995203  550384864_76045959210  823630151_22344597498_646045   
429133    9996877   83108209_22162646095  713500557_31297619237_860300   
256562    9997801   500453195_9636586711  829652786_55390211114_665695   
119534    9998346  292425655_16607136645  189177304_69869411700_783154   
406815    9998447  747535871_38029188908  309540451_47388672022_606106   

       process_step            date_time  
631356        start  2017-06-02 22:32:08  
631354        start  2017-06-02 22:32:28  
639519        start  2017-05-06 08:37:14  
639522        start  2017-05-06 08:36:40  
639530        start  2017-05-06 08:29:00  
...             ...                  ...  
638238        start  2017-06-11 12:23:43  
429133        start  2017-05-03 18:08:32  
256562        start  2017-04-28 10:59:18  
119534       step_3  2017-03-29 15:37:28  
406815        start  2017-06-03 19:17:15  

[10764 rows x 5 columns]

In [25]:
df_footprints.shape

(755405, 5)

In [27]:
df_footprints[df_footprints["client_id"] == 1531]

client_id             visitor_id                      visit_id  \
631353       1531  934069404_34543643308  110350144_67985219545_904677   
631354       1531  934069404_34543643308  110350144_67985219545_904677   
631355       1531  934069404_34543643308  110350144_67985219545_904677   
631356       1531  934069404_34543643308  110350144_67985219545_904677   
631357       1531  934069404_34543643308  110350144_67985219545_904677   
631358       1531  934069404_34543643308  110350144_67985219545_904677   
631359       1531  934069404_34543643308  110350144_67985219545_904677   
631360       1531  934069404_34543643308  110350144_67985219545_904677   
631361       1531  934069404_34543643308  110350144_67985219545_904677   

       process_step            date_time  
631353        start  2017-06-02 22:32:28  
631354        start  2017-06-02 22:32:28  
631355        start  2017-06-02 22:32:08  
631356        start  2017-06-02 22:32:08  
631357      confirm  2017-06-02 22:25:26  
631358       step_3  2017-06-02 22:25:04  
631359       step_2  2017-06-02 22:23:01  
631360       step_1  2017-06-02 22:22:50  
631361        start  2017-06-02 22:22:44

In [38]:
# join with final_experiment

In [35]:
df_footprints.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 755405 entries, 0 to 755404
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   client_id     755405 non-null  int64         
 1   visitor_id    755405 non-null  object        
 2   visit_id      755405 non-null  object        
 3   process_step  755405 non-null  object        
 4   date_time     755405 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 28.8+ MB


In [9]:
df_footprints["date_time"] = pd.to_datetime(df_footprints["date_time"])

In [10]:
df_footprints["date_time"]

0        2017-04-17 15:27:07
1        2017-04-17 15:26:51
2        2017-04-17 15:19:22
3        2017-04-17 15:19:13
4        2017-04-17 15:18:04
                 ...        
755400   2017-05-24 18:46:10
755401   2017-05-24 18:45:29
755402   2017-05-24 18:44:51
755403   2017-05-24 18:44:34
755404   2017-06-03 12:17:09
Name: date_time, Length: 755405, dtype: datetime64[ns]

# Cleaning steps done: 
## Data Types
 Change data type for the column date_time from object to datetime
## Duplicates 
there is 10764 to discuss 
## Column names are correct 

In [10]:
df_final_experiment

client_id Variation
0        9988021      Test
1        8320017      Test
2        4033851   Control
3        1982004      Test
4        9294070   Control
...          ...       ...
70604    2443347       NaN
70605    8788427       NaN
70606     266828       NaN
70607    1266421       NaN
70608    9895983       NaN

[70609 rows x 2 columns]

In [11]:
df_footprints_cleaned = df_footprints.drop_duplicates()

In [12]:
df_footprints_cleaned[df_footprints_cleaned["client_id"] == 1531]

client_id             visitor_id                      visit_id  \
631353       1531  934069404_34543643308  110350144_67985219545_904677   
631355       1531  934069404_34543643308  110350144_67985219545_904677   
631357       1531  934069404_34543643308  110350144_67985219545_904677   
631358       1531  934069404_34543643308  110350144_67985219545_904677   
631359       1531  934069404_34543643308  110350144_67985219545_904677   
631360       1531  934069404_34543643308  110350144_67985219545_904677   
631361       1531  934069404_34543643308  110350144_67985219545_904677   

       process_step           date_time  
631353        start 2017-06-02 22:32:28  
631355        start 2017-06-02 22:32:08  
631357      confirm 2017-06-02 22:25:26  
631358       step_3 2017-06-02 22:25:04  
631359       step_2 2017-06-02 22:23:01  
631360       step_1 2017-06-02 22:22:50  
631361        start 2017-06-02 22:22:44

In [13]:
# df_footprints_cleaned 
df_final_experiment

client_id Variation
0        9988021      Test
1        8320017      Test
2        4033851   Control
3        1982004      Test
4        9294070   Control
...          ...       ...
70604    2443347       NaN
70605    8788427       NaN
70606     266828       NaN
70607    1266421       NaN
70608    9895983       NaN

[70609 rows x 2 columns]

In [14]:
df_final = df_footprints_cleaned.merge(df_final_experiment, how='left', on='client_id')


In [15]:
df_final

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
744636    9668240   388766751_9038881013   922267647_3096648104_968866   
744637    9668240   388766751_9038881013   922267647_3096648104_968866   
744638    9668240   388766751_9038881013   922267647_3096648104_968866   
744639    9668240   388766751_9038881013   922267647_3096648104_968866   
744640     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step           date_time Variation  
0            step_3 2017-04-17 15:27:07      Test  
1            step_2 2017-04-17 15:26:51      Test  
2            step_3 2017-04-17 15:19:22      Test  
3            step_2 2017-04-17 15:19:13      Test  
4            step_3 2017-04-17 15:18:04      Test  
...             ...                 ...       ...  
744636        start 2017-05-24 18:46:10       NaN  
744637        start 2017-05-24 18:45:29       NaN  
744638       step_1 2017-05-24 18:44:51       NaN  
744639        start 2017-05-24 18:44:34       NaN  
744640        start 2017-06-03 12:17:09       NaN  

[744641 rows x 6 columns]

In [16]:
df_final["Variation"].unique()

array(['Test', 'Control', nan], dtype=object)

In [17]:
df_final_v1 = df_final.dropna(subset=['Variation'])


In [18]:
df_final_v1.shape

(317235, 6)

In [19]:
df_final_v1.columns

Index(['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time',
       'Variation'],
      dtype='object')

In [20]:
df_final_v2 = df_final_v1.rename(columns={
    'Variation': 'treatment_group'
})

In [22]:
df_final_v2.to_csv(config['data']['clean']['footprint_final_client_experiment_Cleaned'], index=False, sep=',')
